content/04-postgresql/13-reading-and-querying-data/04-optimizing-postgresql.mdx

```diff
@@ -0,0 +1,663 @@
+
+Some columns that may help you investigate problematic locks include:
+
+* `locktype`: The type of [lockable object](https://www.postgresql.org/docs/current/monitoring-stats.html#WAIT-EVENT-LOCK-TABLE)
+* `database/relation/page/tuple`: The object ID of the locked item.  For database and relations, these can be cross-referenced in the `pg_database` and `pg_class`.
+* `mode`: The [lock mode](https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-TABLES) that is implemented or requested.
+* `granted`: A boolean representing whether the lock was granted.
+
+## Enable slow query logging
+
+One way to find information about long running queries more easily is to enable slow query logging.  Enabling slow query logging allows PostgreSQL to automatically note any queries that take longer to execute than a given amount of time.  This allows you to gather information about slow queries that are not executing at the moment of your investigation.
+
+### Check if PostgreSQL is already logging slow queries
+
+The first thing you should do is verify the current state of slow query logging.  If slow query logging is already enabled, you don't have to do anything.
+
+You can check if slow query logging is enabled by typing:
+
+```sql
+SELECT * FROM pg_settings WHERE name = 'log_min_duration_statement'\gx
+```
+```
+-[ RECORD 1 ]---+---------------------------------------------------------------------------
+name            | log_min_duration_statement
+setting         | -1
+unit            | ms
+category        | Reporting and Logging / When to Log
+short_desc      | Sets the minimum execution time above which all statements will be logged.
+extra_desc      | Zero prints all queries. -1 turns this feature off.
+context         | superuser
+vartype         | integer
+source          | default
+min_val         | -1
+max_val         | 2147483647
+enumvals        |
+boot_val        | -1
+reset_val       | -1
+sourcefile      |
+sourceline      |
+pending_restart | f
+```
+
+If you check the values of the `short_desc` and `extra_desc` columns you will find the information that allows us to evaluate whether logging is currently enabled.  We can see that slow query logging is currently *not* enabled because the `setting` column is currently set to `-1`.
+
+Now that you know the current state, you can change it as necessary.
+
+### Configure PostgreSQL to log slow queries
+
+Before we move on, it is important to note that while slow query logging is incredibly useful, it can potentially have an additional performance impact.  PostgreSQL must perform additional operations to time each query and to record the results to a log.  This can impact performance and fill up hard drive space unexpectedly.
+
+It may not be a good idea to log slow queries at all times.  Instead, enable the functionality when you are actively investigating an issue and disable it when you are finished.
```

Comment:
Great call out here. Demonstrates the balancing act of optimization. Turn it on to find optimizations, but if you leave it on too often, it eventually hinders optimization 😄